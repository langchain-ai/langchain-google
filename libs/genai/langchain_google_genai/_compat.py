"""Go from v1 content blocks to generativelanguage_v1beta format."""

import json
from typing import Any, Optional, cast

from langchain_core.messages import content as types


def translate_citations_to_grounding_metadata(
    citations: list[types.Citation], web_search_queries: Optional[list[str]] = None
) -> dict[str, Any]:
    """Translate LangChain Citations to Google AI grounding metadata format.

    Args:
        citations: List of Citation content blocks.
        web_search_queries: Optional list of search queries that generated
            the grounding data.

    Returns:
        Google AI grounding metadata dictionary.

    Example:
        >>> citations = [
        ...     create_citation(
        ...         url="https://uefa.com/euro2024",
        ...         title="UEFA Euro 2024 Results",
        ...         start_index=0,
        ...         end_index=47,
        ...         cited_text="Spain won the UEFA Euro 2024 championship",
        ...     )
        ... ]
        >>> metadata = translate_citations_to_grounding_metadata(citations)
        >>> len(metadata["groundingChunks"])
        1
        >>> metadata["groundingChunks"][0]["web"]["uri"]
        'https://uefa.com/euro2024'
    """
    if not citations:
        return {}

    # Group citations by text segment (start_index, end_index, cited_text)
    segment_to_citations: dict[
        tuple[Optional[int], Optional[int], Optional[str]], list[types.Citation]
    ] = {}

    for citation in citations:
        key = (
            citation.get("start_index"),
            citation.get("end_index"),
            citation.get("cited_text"),
        )
        if key not in segment_to_citations:
            segment_to_citations[key] = []
        segment_to_citations[key].append(citation)

    # Build grounding chunks from unique URLs
    url_to_chunk_index: dict[str, int] = {}
    grounding_chunks: list[dict[str, Any]] = []

    for citation in citations:
        url = citation.get("url")
        if url and url not in url_to_chunk_index:
            url_to_chunk_index[url] = len(grounding_chunks)
            grounding_chunks.append(
                {"web": {"uri": url, "title": citation.get("title", "")}}
            )

    # Build grounding supports
    grounding_supports: list[dict[str, Any]] = []

    for (
        start_index,
        end_index,
        cited_text,
    ), citations_group in segment_to_citations.items():
        if start_index is not None and end_index is not None and cited_text:
            chunk_indices = []
            confidence_scores = []

            for citation in citations_group:
                url = citation.get("url")
                if url and url in url_to_chunk_index:
                    chunk_indices.append(url_to_chunk_index[url])

                    # Extract confidence scores from extras if available
                    extras = citation.get("extras", {})
                    google_metadata = extras.get("google_ai_metadata", {})
                    scores = google_metadata.get("confidence_scores", [])
                    confidence_scores.extend(scores)

            support = {
                "segment": {
                    "startIndex": start_index,
                    "endIndex": end_index,
                    "text": cited_text,
                },
                "groundingChunkIndices": chunk_indices,
            }

            if confidence_scores:
                support["confidenceScores"] = confidence_scores

            grounding_supports.append(support)

    # Extract search queries from extras if not provided
    if web_search_queries is None:
        web_search_queries = []
        for citation in citations:
            extras = citation.get("extras", {})
            google_metadata = extras.get("google_ai_metadata", {})
            queries = google_metadata.get("web_search_queries", [])
            web_search_queries.extend(queries)
        # Remove duplicates while preserving order
        web_search_queries = list(dict.fromkeys(web_search_queries))

    return {
        "webSearchQueries": web_search_queries,
        "groundingChunks": grounding_chunks,
        "groundingSupports": grounding_supports,
    }


def _convert_from_v1_to_generativelanguage_v1beta(
    content: list[types.ContentBlock], model_provider: str | None
) -> list[dict[str, Any]]:
    """Convert v1 content blocks to `google.ai.generativelanguage_v1beta.types.Content`.

    Args:
        content: List of v1 `ContentBlock` objects.
        model_provider: The model provider name that generated the v1 content.

    Returns:
        List of dictionaries in `google.ai.generativelanguage_v1beta.types.Content`
        format, ready to be sent to the API.
    """
    new_content: list = []
    for block in content:
        if not isinstance(block, dict) or "type" not in block:
            continue

        block_dict = dict(block)  # (For typing)

        # TextContentBlock
        if block_dict["type"] == "text":
            new_block = {"text": block_dict.get("text", "")}
            new_content.append(new_block)
            # Citations are only handled on output. Can't pass them back :/

        # ReasoningContentBlock -> thinking
        elif block_dict["type"] == "reasoning" and model_provider == "google_genai":
            # Google requires passing back the thought_signature when available.
            # Signatures are only provided when function calling is enabled.
            if "extras" in block_dict and isinstance(block_dict["extras"], dict):
                extras = block_dict["extras"]
                if "signature" in extras:
                    new_block = {
                        "thought": True,
                        "text": block_dict.get("reasoning", ""),
                        "thought_signature": extras["signature"],
                    }
                    new_content.append(new_block)
                # else: skip reasoning blocks without signatures
                # TODO: log a warning?
            # else: skip reasoning blocks without extras
            # TODO: log a warning?

        # ImageContentBlock
        elif block_dict["type"] == "image":
            if base64 := block_dict.get("base64"):
                new_block = {
                    "inline_data": {
                        "mime_type": block_dict.get("mime_type", "image/jpeg"),
                        "data": base64.encode("utf-8")
                        if isinstance(base64, str)
                        else base64,
                    }
                }
                new_content.append(new_block)
            elif url := block_dict.get("url") and model_provider == "google_genai":
                # Google file service
                new_block = {
                    "file_data": {
                        "mime_type": block_dict.get("mime_type", "image/jpeg"),
                        "file_uri": block_dict[str(url)],
                    }
                }
                new_content.append(new_block)

        # TODO: AudioContentBlock -> audio once models support passing back in

        # FileContentBlock (documents)
        elif block_dict["type"] == "file":
            if base64 := block_dict.get("base64"):
                new_block = {
                    "inline_data": {
                        "mime_type": block_dict.get(
                            "mime_type", "application/octet-stream"
                        ),
                        "data": base64.encode("utf-8")
                        if isinstance(base64, str)
                        else base64,
                    }
                }
                new_content.append(new_block)
            elif url := block_dict.get("url") and model_provider == "google_genai":
                # Google file service
                new_block = {
                    "file_data": {
                        "mime_type": block_dict.get(
                            "mime_type", "application/octet-stream"
                        ),
                        "file_uri": block_dict[str(url)],
                    }
                }
                new_content.append(new_block)

        # ToolCall -> FunctionCall
        elif block_dict["type"] == "tool_call":
            function_call = {
                "function_call": {
                    "name": block_dict.get("name", ""),
                    "args": block_dict.get("args", {}),
                }
            }
            new_content.append(function_call)

        # ToolCallChunk -> FunctionCall
        elif block_dict["type"] == "tool_call_chunk":
            try:
                args_str = block_dict.get("args") or "{}"
                input_ = json.loads(args_str) if isinstance(args_str, str) else args_str
            except json.JSONDecodeError:
                input_ = {}

            function_call = {
                "function_call": {
                    "name": block_dict.get("name", "no_tool_name_present"),
                    "args": input_,
                }
            }
            new_content.append(function_call)

        elif block_dict["type"] == "server_tool_call":
            if block_dict.get("name") == "code_interpreter":
                # LangChain v0 format
                args = cast(dict, block_dict.get("args", {}))
                executable_code = {
                    "type": "executable_code",
                    "executable_code": args.get("code", ""),
                    "language": args.get("language", ""),
                    "id": block_dict.get("id", ""),
                }
                # Google generativelanguage format
                new_content.append(
                    {
                        "executable_code": {
                            "language": executable_code["language"],
                            "code": executable_code["executable_code"],
                        }
                    }
                )

        elif block_dict["type"] == "server_tool_result":
            extras = cast(dict, block_dict.get("extras", {}))
            if extras.get("block_type") == "code_execution_result":
                # LangChain v0 format
                code_execution_result = {
                    "type": "code_execution_result",
                    "code_execution_result": block_dict.get("output", ""),
                    "outcome": extras.get("outcome", ""),
                    "tool_call_id": block_dict.get("tool_call_id", ""),
                }
                # Google generativelanguage format
                new_content.append(
                    {
                        "code_execution_result": {
                            "outcome": code_execution_result["outcome"],
                            "output": code_execution_result["code_execution_result"],
                        }
                    }
                )

        elif block_dict["type"] == "non_standard":
            new_content.append(block_dict["value"])

    return new_content
