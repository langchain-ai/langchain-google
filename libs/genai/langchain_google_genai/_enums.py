import google.ai.generativelanguage_v1beta as genai

HarmBlockThreshold = genai.SafetySetting.HarmBlockThreshold
HarmCategory = genai.HarmCategory

__all__ = ["HarmBlockThreshold", "HarmCategory"]
