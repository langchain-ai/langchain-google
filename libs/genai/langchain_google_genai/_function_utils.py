from __future__ import annotations

from typing import (
    Any,
    Dict,
    List,
    Literal,
    Optional,
    Sequence,
    Type,
    TypedDict,
    Union,
)

import google.ai.generativelanguage as glm
from google.generativeai.types import Tool as GoogleTool  # type: ignore[import]
from google.generativeai.types.content_types import (  # type: ignore[import]
    FunctionCallingConfigType,
    FunctionDeclarationType,
    ToolDict,
    ToolType,
)
from langchain_core.pydantic_v1 import BaseModel
from langchain_core.tools import BaseTool
from langchain_core.tools import tool as callable_as_lc_tool
from langchain_core.utils.json_schema import dereference_refs

TYPE_ENUM = {
    "string": glm.Type.STRING,
    "number": glm.Type.NUMBER,
    "integer": glm.Type.INTEGER,
    "boolean": glm.Type.BOOLEAN,
    "array": glm.Type.ARRAY,
    "object": glm.Type.OBJECT,
}

TYPE_ENUM_REVERSE = {v: k for k, v in TYPE_ENUM.items()}


def convert_to_genai_function_declarations(
    tool: Union[
        GoogleTool, ToolDict, FunctionDeclarationType, Sequence[FunctionDeclarationType]
    ],
) -> ToolType:
    """Convert any tool-like object to a ToolType.

    https://github.com/google-gemini/generative-ai-python/blob/668695ebe3e9de496a36eeb95cb2ed2faba9b939/google/generativeai/types/content_types.py#L574
    """
    if isinstance(tool, GoogleTool):
        return tool
    # check whether a dict is supported by glm, otherwise we parse it explicitly
    if isinstance(tool, dict):
        first_function_declaration = tool.get("function_declarations", [None])[0]
        if isinstance(first_function_declaration, glm.FunctionDeclaration):
            return tool
        schema = None
        try:
            schema = first_function_declaration.parameters
        except AttributeError:
            pass
        if schema is None:
            schema = first_function_declaration.get("parameters")
        if schema is None or isinstance(schema, glm.Schema):
            return tool
        return glm.Tool(
            function_declarations=[
                _convert_to_genai_function(fc) for fc in tool["function_declarations"]
            ],
        )
    elif isinstance(tool, type) and issubclass(tool, BaseModel):
        return glm.Tool(function_declarations=[_convert_to_genai_function(tool)])
    elif callable(tool):
        return _convert_tool_to_genai_function(callable_as_lc_tool()(tool))
    elif isinstance(tool, list):
        return glm.Tool(
            function_declarations=[_convert_to_genai_function(fc) for fc in tool]
        )
    return glm.Tool(function_declarations=[_convert_to_genai_function(tool)])


def tool_to_dict(tool: Union[glm.Tool, GoogleTool]) -> ToolDict:
    if isinstance(tool, GoogleTool):
        tool = tool._proto
    function_declarations = []
    for function_declaration_proto in tool.function_declarations:
        properties: Dict[str, Any] = {}
        for property in function_declaration_proto.parameters.properties:
            property_type = function_declaration_proto.parameters.properties[
                property
            ].type
            property_dict = {"type": TYPE_ENUM_REVERSE[property_type]}
            property_description = function_declaration_proto.parameters.properties[
                property
            ].description
            if property_description:
                property_dict["description"] = property_description
            properties[property] = property_dict
        function_declaration = {
            "name": function_declaration_proto.name,
            "description": function_declaration_proto.description,
            "parameters": {"type": "object", "properties": properties},
        }
        if function_declaration_proto.parameters.required:
            function_declaration["parameters"][  # type: ignore[index]
                "required"
            ] = function_declaration_proto.parameters.required
        function_declarations.append(function_declaration)
    return {"function_declarations": function_declarations}


def _convert_to_genai_function(fc: FunctionDeclarationType) -> glm.FunctionDeclaration:
    if isinstance(fc, BaseTool):
        return _convert_tool_to_genai_function(fc)
    elif isinstance(fc, type) and issubclass(fc, BaseModel):
        return _convert_pydantic_to_genai_function(fc)
    elif callable(fc):
        return _convert_tool_to_genai_function(callable_as_lc_tool()(fc))
    elif isinstance(fc, dict):
        return glm.FunctionDeclaration(
            name=fc["name"],
            description=fc.get("description"),
            parameters={
                "properties": {
                    k: {
                        "type_": TYPE_ENUM[v["type"]],
                        "description": v.get("description"),
                    }
                    for k, v in fc["parameters"]["properties"].items()
                },
                "required": fc["parameters"].get("required", []),
                "type_": TYPE_ENUM[fc["parameters"]["type"]],
            },
        )
    else:
        raise ValueError(f"Unsupported function call type {fc}")


def _convert_tool_to_genai_function(tool: BaseTool) -> glm.FunctionDeclaration:
    if tool.args_schema:
        schema = dereference_refs(tool.args_schema.schema())
        schema.pop("definitions", None)
        return glm.FunctionDeclaration(
            name=tool.name or schema["title"],
            description=tool.description or schema["description"],
            parameters={
                "properties": {
                    k: {
                        "type_": TYPE_ENUM[v["type"]],
                        "description": v.get("description"),
                    }
                    for k, v in schema["properties"].items()
                },
                "required": schema.get("required", []),
                "type_": TYPE_ENUM[schema["type"]],
            },
        )
    else:
        return glm.FunctionDeclaration(
            name=tool.name,
            description=tool.description,
            parameters={
                "properties": {
                    "__arg1": {"type_": TYPE_ENUM["string"]},
                },
                "required": ["__arg1"],
                "type_": TYPE_ENUM["object"],
            },
        )


def _convert_pydantic_to_genai_function(
    pydantic_model: Type[BaseModel],
) -> glm.FunctionDeclaration:
    schema = dereference_refs(pydantic_model.schema())
    schema.pop("definitions", None)
    return glm.FunctionDeclaration(
        name=schema["title"],
        description=schema.get("description", ""),
        parameters={
            "properties": {
                k: {
                    "type_": TYPE_ENUM[v["type"]],
                    "description": v.get("description"),
                }
                for k, v in schema["properties"].items()
            },
            "required": schema["required"],
            "type_": TYPE_ENUM[schema["type"]],
        },
    )


_ToolChoiceType = Union[
    dict, List[str], str, Literal["auto", "none", "any"], Literal[True]
]


class _ToolConfigDict(TypedDict):
    function_calling_config: FunctionCallingConfigType


def _tool_choice_to_tool_config(
    tool_choice: _ToolChoiceType,
    all_names: List[str],
) -> _ToolConfigDict:
    allowed_function_names: Optional[List[str]] = None
    if tool_choice is True or tool_choice == "any":
        mode = "any"
        allowed_function_names = all_names
    elif tool_choice == "auto":
        mode = "auto"
    elif tool_choice == "none":
        mode = "none"
    elif isinstance(tool_choice, str):
        mode = "any"
        allowed_function_names = [tool_choice]
    elif isinstance(tool_choice, list):
        mode = "any"
        allowed_function_names = tool_choice
    elif isinstance(tool_choice, dict):
        if "mode" in tool_choice:
            mode = tool_choice["mode"]
            allowed_function_names = tool_choice.get("allowed_function_names")
        elif "function_calling_config" in tool_choice:
            mode = tool_choice["function_calling_config"]["mode"]
            allowed_function_names = tool_choice["function_calling_config"].get(
                "allowed_function_names"
            )
        else:
            raise ValueError(
                f"Unrecognized tool choice format:\n\n{tool_choice=}\n\nShould match "
                f"Google GenerativeAI ToolConfig or FunctionCallingConfig format."
            )
    else:
        raise ValueError(f"Unrecognized tool choice format:\n\n{tool_choice=}")
    return _ToolConfigDict(
        function_calling_config={
            "mode": mode,
            "allowed_function_names": allowed_function_names,
        }
    )
