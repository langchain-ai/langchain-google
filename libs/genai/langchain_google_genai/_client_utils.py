import asyncio
from functools import lru_cache
from typing import Optional
from weakref import WeakKeyDictionary

from google.ai.generativelanguage_v1beta import (
    GenerativeServiceAsyncClient as v1betaGenerativeServiceAsyncClient,
)
from google.ai.generativelanguage_v1beta import (
    GenerativeServiceClient as v1betaGenerativeServiceClient,
)

from langchain_google_genai._common import get_client_info

from . import _genai_extension as genaix


# Cache sync client
@lru_cache
def _get_sync_client(
    *,
    api_key: Optional[str],
    model: str,
    transport: str,
) -> v1betaGenerativeServiceClient:
    """Return a shared sync client."""
    client_info = get_client_info(f"ChatGoogleGenerativeAI:{model}")
    return genaix.build_generative_service(
        api_key=api_key,
        client_info=client_info,
        client_options=None,
        transport=transport,
    )


# Cache async client - must store caches per event loop
_client_caches: WeakKeyDictionary = WeakKeyDictionary()


def _create_async_client(
    *,
    api_key: Optional[str],
    model: str,
    transport: Optional[str],
) -> v1betaGenerativeServiceAsyncClient:
    """Create a new async client."""
    # async clients don't support "rest" transport
    # https://github.com/googleapis/gapic-generator-python/issues/1962
    if transport == "rest":
        transport = "grpc_asyncio"
    return genaix.build_generative_async_service(
        credentials=None,
        api_key=api_key,
        client_info=get_client_info(f"ChatGoogleGenerativeAI:{model}"),
        client_options=None,
        transport=transport,
    )


def _get_async_client(
    *,
    api_key: Optional[str],
    model: str,
    transport: Optional[str],
) -> v1betaGenerativeServiceAsyncClient:
    """Return a shared async client per event loop."""
    try:
        loop = asyncio.get_running_loop()
    except RuntimeError:
        # If no event loop is running, don't cache
        return _create_async_client(
            api_key=api_key,
            model=model,
            transport=transport,
        )

    # Get or create cache for this event loop
    if loop not in _client_caches:
        _client_caches[loop] = {}

    cache_key = (api_key, model, transport)

    if cache_key not in _client_caches[loop]:
        _client_caches[loop][cache_key] = _create_async_client(
            api_key=api_key,
            model=model,
            transport=transport,
        )

    return _client_caches[loop][cache_key]
