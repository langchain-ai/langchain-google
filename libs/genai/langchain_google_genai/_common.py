from importlib import metadata
from typing import Optional, Tuple, TypedDict

from google.api_core.gapic_v1.client_info import ClientInfo

from langchain_google_genai._enums import HarmBlockThreshold, HarmCategory


class GoogleGenerativeAIError(Exception):
    """
    Custom exception class for errors associated with the `Google GenAI` API.
    """


def get_user_agent(module: Optional[str] = None) -> Tuple[str, str]:
    r"""Returns a custom user agent header.

    Args:
        module (Optional[str]):
            Optional. The module for a custom user agent header.
    Returns:
        Tuple[str, str]
    """
    try:
        langchain_version = metadata.version("langchain-google-genai")
    except metadata.PackageNotFoundError:
        langchain_version = "0.0.0"
    client_library_version = (
        f"{langchain_version}-{module}" if module else langchain_version
    )
    return client_library_version, f"langchain-google-genai/{client_library_version}"


def get_client_info(module: Optional[str] = None) -> "ClientInfo":
    r"""Returns a client info object with a custom user agent header.

    Args:
        module (Optional[str]):
            Optional. The module for a custom user agent header.
    Returns:
        google.api_core.gapic_v1.client_info.ClientInfo
    """
    client_library_version, user_agent = get_user_agent(module)
    return ClientInfo(
        client_library_version=client_library_version,
        user_agent=user_agent,
    )


class SafetySettingDict(TypedDict):
    category: HarmCategory
    threshold: HarmBlockThreshold
