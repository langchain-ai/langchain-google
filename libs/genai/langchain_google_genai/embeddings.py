from typing import Any, Dict, List, Optional

# TODO: remove ignore once the google package is published with types
from google.ai.generativelanguage_v1beta.types import (
    BatchEmbedContentsRequest,
    EmbedContentRequest,
)
from langchain_core.embeddings import Embeddings
from langchain_core.pydantic_v1 import BaseModel, Field, SecretStr, root_validator
from langchain_core.utils import get_from_dict_or_env

from langchain_google_genai._common import (
    GoogleGenerativeAIError,
    get_client_info,
)
from langchain_google_genai._genai_extension import build_generative_service


class GoogleGenerativeAIEmbeddings(BaseModel, Embeddings):
    """`Google Generative AI Embeddings`.

    To use, you must have either:

        1. The ``GOOGLE_API_KEY``` environment variable set with your API key, or
        2. Pass your API key using the google_api_key kwarg to the ChatGoogle
           constructor.

    Example:
        .. code-block:: python

            from langchain_google_genai import GoogleGenerativeAIEmbeddings

            embeddings = GoogleGenerativeAIEmbeddings(model="models/embedding-001")
            embeddings.embed_query("What's our Q1 revenue?")
    """

    client: Any  #: :meta private:
    model: str = Field(
        ...,
        description="The name of the embedding model to use. "
        "Example: models/embedding-001",
    )
    task_type: Optional[str] = Field(
        None,
        description="The task type. Valid options include: "
        "task_type_unspecified, retrieval_query, retrieval_document, "
        "semantic_similarity, classification, and clustering",
    )
    google_api_key: Optional[SecretStr] = Field(
        None,
        description="The Google API key to use. If not provided, "
        "the GOOGLE_API_KEY environment variable will be used.",
    )
    credentials: Any = Field(
        default=None,
        exclude=True,
        description="The default custom credentials "
        "(google.auth.credentials.Credentials) to use when making API calls. If not "
        "provided, credentials will be ascertained from the GOOGLE_API_KEY envvar",
    )
    client_options: Optional[Dict] = Field(
        None,
        description=(
            "A dictionary of client options to pass to the Google API client, "
            "such as `api_endpoint`."
        ),
    )
    transport: Optional[str] = Field(
        None,
        description="A string, one of: [`rest`, `grpc`, `grpc_asyncio`].",
    )
    request_options: Optional[Dict] = Field(
        None,
        description="A dictionary of request options to pass to the Google API client."
        "Example: `{'timeout': 10}`",
    )

    @root_validator()
    def validate_environment(cls, values: Dict) -> Dict:
        """Validates params and passes them to google-generativeai package."""
        google_api_key = get_from_dict_or_env(
            values, "google_api_key", "GOOGLE_API_KEY"
        )
        client_info = get_client_info("GoogleGenerativeAIEmbeddings")

        values["client"] = build_generative_service(
            credentials=values.get("credentials"),
            api_key=google_api_key,
            client_info=client_info,
            client_options=values.get("client_options"),
        )
        return values

    def _prepare_request(
        self,
        text: str,
        task_type: Optional[str] = None,
        title: Optional[str] = None,
        output_dimensionality: Optional[int] = None,
    ) -> EmbedContentRequest:
        task_type = self.task_type or task_type or "RETRIEVAL_DOCUMENT"
        # https://ai.google.dev/api/rest/v1/models/batchEmbedContents#EmbedContentRequest
        request = EmbedContentRequest(
            content={"parts": [{"text": text}]},
            model=self.model,
            task_type=task_type.upper(),
            title=title,
            output_dimensionality=output_dimensionality,
        )
        return request

    def embed_documents(
        self,
        texts: List[str],
        task_type: Optional[str] = None,
        titles: Optional[List[str]] = None,
        output_dimensionality: Optional[int] = None,
    ) -> List[List[float]]:
        """Embed a list of strings. Vertex AI currently
        sets a max batch size of 5 strings.

        Args:
            texts: List[str] The list of strings to embed.
            batch_size: [int] The batch size of embeddings to send to the model
            task_type: task_type (https://ai.google.dev/api/rest/v1/TaskType)
            titles: An optional list of titles for texts provided.
            Only applicable when TaskType is RETRIEVAL_DOCUMENT.
            output_dimensionality: Optional reduced dimension for the output embedding.
            https://ai.google.dev/api/rest/v1/models/batchEmbedContents#EmbedContentRequest

        Returns:
            List of embeddings, one for each text.
        """
        titles = titles if titles else [None] * len(texts)  # type: ignore[list-item]
        requests = [
            self._prepare_request(
                text=text,
                task_type=task_type,
                title=title,
                output_dimensionality=output_dimensionality,
            )
            for text, title in zip(texts, titles)
        ]

        try:
            result = self.client.batch_embed_contents(
                BatchEmbedContentsRequest(requests=requests, model=self.model)
            )
        except Exception as e:
            raise GoogleGenerativeAIError(f"Error embedding content: {e}") from e
        return [e.values for e in result.embeddings]

    def embed_query(
        self,
        text: str,
        task_type: Optional[str] = None,
        title: Optional[str] = None,
        output_dimensionality: Optional[int] = None,
    ) -> List[float]:
        """Embed a text.

        Args:
            text: The text to embed.
            task_type: task_type (https://ai.google.dev/api/rest/v1/TaskType)
            title: An optional title for the text.
            Only applicable when TaskType is RETRIEVAL_DOCUMENT.
            output_dimensionality: Optional reduced dimension for the output embedding.
            https://ai.google.dev/api/rest/v1/models/batchEmbedContents#EmbedContentRequest

        Returns:
            Embedding for the text.
        """
        task_type = self.task_type or "RETRIEVAL_QUERY"
        return self.embed_documents(
            [text],
            task_type=task_type,
            titles=[title] if title else None,
            output_dimensionality=output_dimensionality,
        )[0]
