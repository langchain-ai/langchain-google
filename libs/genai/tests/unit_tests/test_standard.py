from typing import Tuple, Type

from langchain_core.language_models import BaseChatModel
from langchain_standard_tests.unit_tests import ChatModelUnitTests

from langchain_google_genai import ChatGoogleGenerativeAI


class TestGeminiAIStandard(ChatModelUnitTests):
    @property
    def chat_model_class(self) -> Type[BaseChatModel]:
        return ChatGoogleGenerativeAI

    @property
    def chat_model_params(self) -> dict:
        return {"model": "models/gemini-1.0-pro-001"}

    @property
    def init_from_env_params(self) -> Tuple[dict, dict, dict]:
        return (
            {"GOOGLE_API_KEY": "api_key"},
            self.chat_model_params,
            {"google_api_key": "api_key"},
        )


class TestGemini_15_AIStandard(ChatModelUnitTests):
    @property
    def chat_model_class(self) -> Type[BaseChatModel]:
        return ChatGoogleGenerativeAI

    @property
    def chat_model_params(self) -> dict:
        return {"model": "models/gemini-1.5-pro-001"}

    @property
    def init_from_env_params(self) -> Tuple[dict, dict, dict]:
        return (
            {"GOOGLE_API_KEY": "api_key"},
            self.chat_model_params,
            {"google_api_key": "api_key"},
        )
