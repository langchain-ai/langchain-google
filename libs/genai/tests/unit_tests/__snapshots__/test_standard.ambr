# serializer version: 1
# name: TestGeminiAIStandard.test_serdes[serialized]
  dict({
    'id': list([
      'langchain_google_genai',
      'chat_models',
      'ChatGoogleGenerativeAI',
    ]),
    'kwargs': dict({
      'default_metadata': list([
      ]),
      'google_api_key': dict({
        'id': list([
          'GOOGLE_API_KEY',
        ]),
        'lc': 1,
        'type': 'secret',
      }),
      'max_retries': 2,
      'model': 'models/gemini-1.0-pro-001',
      'n': 1,
      'temperature': 0.0,
      'timeout': 60.0,
    }),
    'lc': 1,
    'name': 'ChatGoogleGenerativeAI',
    'type': 'constructor',
  })
# ---
# name: TestGemini_15_AIStandard.test_serdes[serialized]
  dict({
    'id': list([
      'langchain_google_genai',
      'chat_models',
      'ChatGoogleGenerativeAI',
    ]),
    'kwargs': dict({
      'default_metadata': list([
      ]),
      'google_api_key': dict({
        'id': list([
          'GOOGLE_API_KEY',
        ]),
        'lc': 1,
        'type': 'secret',
      }),
      'max_retries': 2,
      'model': 'models/gemini-1.5-pro-001',
      'n': 1,
      'temperature': 0.0,
      'timeout': 60.0,
    }),
    'lc': 1,
    'name': 'ChatGoogleGenerativeAI',
    'type': 'constructor',
  })
# ---
